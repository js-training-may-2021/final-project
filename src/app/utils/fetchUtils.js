import axios from 'axios';
import _ from 'lodash';

const isDev = process.env.NODE_ENV === 'development';
const pathApi = isDev ? 'http://localhost:3000' : '/api/v1';
export const limit = 20;

const fetchPokemons = (page, query, attr, order) => axios.get(`${pathApi}/pokemons?q=${query}&_page=${page}&_limit=${limit}&_sort=${attr}&_order=${order}`);
const normalizedQuery = (query) => _.trimStart(query, '#');

export const mappingFetchPokemons = {
  idAsc: (page, query) => fetchPokemons(page, normalizedQuery(query), 'id', 'asc'),
  idDesc: (page, query) => fetchPokemons(page, normalizedQuery(query), 'id', 'desc'),
  nameAsc: (page, query) => fetchPokemons(page, normalizedQuery(query), 'name', 'asc'),
  nameDesc: (page, query) => fetchPokemons(page, normalizedQuery(query), 'name', 'desc'),
};

export const fetchPokemonById = (id) => axios.get(`${pathApi}/pokemons?id=${id}`);

export const fetchData = () => (
  axios({
    method: 'post',
    url: '/api/v1/data',
    data: { token: localStorage.token },
    timeout: 4000,
  })
);

export const fetchCatch = (pokemon) => axios({
  method: 'post',
  url: '/api/v1/catchPokemon',
  data: { ...pokemon, token: localStorage.token },
  timeout: 4000,
});

export const fetchLogin = (data) => axios({
  method: 'post',
  url: '/api/v1/login',
  data,
  timeout: 4000,
});
