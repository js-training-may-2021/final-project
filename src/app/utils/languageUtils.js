import { useTranslation } from 'react-i18next';

export const setDefaultLanguage = () => {
  const defaultLanguage = 'ru';

  if (localStorage.getItem('language') === null) {
    localStorage.setItem('language', defaultLanguage);
  }
};

export const getCurrentLanguage = () => localStorage.getItem('language');

export const toggleLanguage = () => {
  const currentLanguage = localStorage.getItem('language');
  const newLanguage = currentLanguage === 'ru' ? 'en' : 'ru';
  localStorage.setItem('language', newLanguage);

  return newLanguage;
};

export const loc = (text) => {
  const { t } = useTranslation();
  return t(text);
};
