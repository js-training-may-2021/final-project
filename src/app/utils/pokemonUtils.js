import _ from 'lodash';

export const isFound = ({ name, id }, searchText) => {
  const searchBy = (attr, text) => String(attr).includes(text);
  const normalizedText = _.trimStart(searchText.trim().toLowerCase(), '#');
  return searchBy(name, normalizedText) || searchBy(id, normalizedText);
};

export const getNextId = (currentId) => {
  const nextId = currentId + 1 === 803 ? 10001 : currentId + 1;
  return nextId === 10148 ? 1 : nextId;
};

export const getPreviousId = (currentId) => {
  const prevId = currentId - 1 === 0 ? 10147 : currentId - 1;
  return prevId === 10000 ? 802 : prevId;
};

export const mappingSortPokemons = {
  idAsc: (array) => _.orderBy(array, 'id', 'asc'),
  idDesc: (array) => _.orderBy(array, 'id', 'desc'),
  nameAsc: (array) => _.orderBy(array, 'name', 'asc'),
  nameDesc: (array) => _.orderBy(array, 'name', 'desc'),
};
