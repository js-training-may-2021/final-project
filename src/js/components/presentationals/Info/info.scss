@import '../../../../styles/variables';

.info {
  min-width: $mobileWidth;
  padding: 40px;

  &__text {
    font-weight: 700;
  }

  &__data {
    font-weight: 400;
  }

  @media (max-width: $mobileWidth) {
    min-width: 300px;
  }
}