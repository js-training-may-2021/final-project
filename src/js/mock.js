export const pokemons = [
  {
    name: 'bulbasaur',
    id: 1,
    isCaught: false,
    catchDate: null,
  },
  {
    name: 'ivysaur',
    id: 2,
    isCaught: true,
    catchDate: '23 June 2021',
  },
  {
    name: 'venusaur',
    id: 3,
    isCaught: true,
    catchDate: '23 June 2021'
  },
  {
    name: 'charmander',
    id: 4,
    isCaught: false,
    catchDate: null,
  },
  {
    name: 'charmeleon',
    id: 5,
    isCaught: true,
    catchDate: '23 June 2021'
  },
  {
    name: 'charizard',
    id: 6,
    isCaught: true,
    catchDate: '23 June 2021'
  },
  {
    name: 'squirtle',
    id: 7,
    isCaught: false,
    catchDate: null,
  },
  {
    name: 'wartortle',
    id: 8,
    isCaught: true,
    catchDate: '23 June 2021'
  },
  {
    name: 'blastoise',
    id: 9,
    isCaught: true,
    catchDate: '23 June 2021'
  },
  {
    name: 'caterpie',
    id: 10,
    isCaught: false,
    catchDate: null,
  },
  {
    name: 'metapod',
    id: 11,
    isCaught: false,
    catchDate: null,
  }];  

export const uncaughtPokemon = pokemons[0];
export const caughtPokemon = pokemons[1];
