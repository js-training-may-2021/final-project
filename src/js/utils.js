import {createMemoryHistory} from 'history';
import configureStore from 'redux-mock-store';

export const POKEMONS_PER_PAGE = 10;
const START_PAGE = 1;

let currentPage = START_PAGE;

export const getCurrentPage = () => {
	return currentPage++;
};

export const AppRoute = {
  MAIN: '/',
  DETAIL: '/detail/:id',
  CAUGHT: '/caught'
};

const monthNames = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];

export const getCatchDate = () => {
  const date = new Date();

  const day = date.getDate();
  const month = monthNames[date.getMonth()];
  const year = date.getFullYear();

  return `${day} ${month} ${year}`
};

export const noop = () => {};
export const history = createMemoryHistory();
const mockStore = configureStore();
export const testStore = mockStore({});
