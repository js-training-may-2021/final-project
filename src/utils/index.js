export const firstLetterToUpperCase = (str) => {
  if (str === undefined) {
    return str;
  } else {
    return str
      .toLowerCase()
      .split(" ")
      .map((s) => s.charAt(0).toUpperCase() + s.substring(1))
      .join(" ");
  }
};

export const formatDate = (dateInput) => {
  return dateInput.toLocaleString("ru-RU");
};

export const padToThree = (number) => number <= 999 ? `00${number}`.slice(-3) : number;

export const handleErrors = (response) => {
  if (!response.ok) {
    throw Error(`${response.status}: ${response.statusText}`);
  }
  return response;
};
