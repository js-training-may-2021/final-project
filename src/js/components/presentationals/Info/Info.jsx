import React, {Fragment} from 'react';
import PropTypes from 'prop-types';
import Button from '../Button/Button.jsx';
import withToggle from '../../hocs/withToggle/withToggle.jsx';
import './info.scss';

const ButtonWrapped = withToggle(Button);

const Info = (props) => {
  const {
    pokemon,
    isToggleChecked,
    onButtonClick,
    onToggleClick,
  } = props;

	const {id, name, catchDate} = pokemon;

  const showCatchDate = () => {
    return (
      <p className="info__text">CAUGHT: <span className="info__data">{catchDate}</span></p>
    );
  };

  return (
    <Fragment>
      <img className="page__img" src={require(`../../../../images/${id}.png`)} alt={name[0].toUpperCase() + name.slice(1)} />
      <article className="page__info info">
        <h1 className="info__title">{name[0].toUpperCase() + name.slice(1)}</h1>
        <p className="info__text">ID: <span className="info__data">{id}</span></p>
        <p className="info__text">STATUS: <span className="info__data">{isToggleChecked ? `Caught` : `Not caught`}</span></p>
        {isToggleChecked ? showCatchDate() : ``}
				<ButtonWrapped
          isToggleChecked={isToggleChecked}
          onClick={() => {
            onButtonClick(pokemon);
            onToggleClick();
          }}
				/>
      </article>
    </Fragment>
  );
}

Info.propTypes = {
	pokemon: PropTypes.shape({
		id: PropTypes.number.isRequired,
		name: PropTypes.string.isRequired,
		isCaught: PropTypes.bool.isRequired,
		catchDate: PropTypes.oneOfType([() => null, PropTypes.instanceOf(Date)]),
	}).isRequired,
  isToggleChecked: PropTypes.bool.isRequired,
  onButtonClick: PropTypes.func.isRequired,
  onToggleClick: PropTypes.func,
}

export default React.memo(Info);
