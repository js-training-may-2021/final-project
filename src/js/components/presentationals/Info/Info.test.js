import React from 'react';
import renderer from 'react-test-renderer';
import Info from './Info.jsx';
import {noop} from '../../../utils.js';
import {uncaughtPokemon, caughtPokemon} from '../../../mock.js';

it(`Render Info of uncaught pokemon`, () => {
  const tree = renderer
    .create(
      <Info
        pokemon={uncaughtPokemon}
        isToggleChecked={uncaughtPokemon.isCaught}
        onButtonClick={noop}
        onToggleClick={noop}
      />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it(`Render Info of caught pokemon`, () => {
  const tree = renderer
    .create(
      <Info
        pokemon={caughtPokemon}
        isToggleChecked={caughtPokemon.isCaught}
        onButtonClick={noop}
        onToggleClick={noop}
      />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
