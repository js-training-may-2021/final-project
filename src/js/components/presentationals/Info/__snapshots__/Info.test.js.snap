// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Info of caught pokemon 1`] = `
Array [
  <img
    alt="Ivysaur"
    className="page__img"
    src="test-file-stub"
  />,
  <article
    className="page__info info"
  >
    <h1
      className="info__title"
    >
      Ivysaur
    </h1>
    <p
      className="info__text"
    >
      ID: 
      <span
        className="info__data"
      >
        2
      </span>
    </p>
    <p
      className="info__text"
    >
      STATUS: 
      <span
        className="info__data"
      >
        Caught
      </span>
    </p>
    <p
      className="info__text"
    >
      CAUGHT: 
      <span
        className="info__data"
      >
        23 June 2021
      </span>
    </p>
    <button
      className="card__button button button--checked"
      onClick={[Function]}
      type="button"
    >
      Release
    </button>
  </article>,
]
`;

exports[`Render Info of uncaught pokemon 1`] = `
Array [
  <img
    alt="Bulbasaur"
    className="page__img"
    src="test-file-stub"
  />,
  <article
    className="page__info info"
  >
    <h1
      className="info__title"
    >
      Bulbasaur
    </h1>
    <p
      className="info__text"
    >
      ID: 
      <span
        className="info__data"
      >
        1
      </span>
    </p>
    <p
      className="info__text"
    >
      STATUS: 
      <span
        className="info__data"
      >
        Not caught
      </span>
    </p>
    <button
      className="card__button button"
      onClick={[Function]}
      type="button"
    >
      Catch!
    </button>
  </article>,
]
`;
